src/app/modules/form/public-form/public-form.controller.ts

```diff
@@ -161,3 +171,141 @@ export const handleRedirect: RequestHandler<
+    FormService.checkFormSubmissionLimitAndDeactivateForm(form),
+  )
+
+  // Early return if form is not public or any error occurred.
+  if (formResult.isErr()) {
+    const { error } = formResult
+    logger.error({
+      message: 'Error retrieving public form',
+      meta: {
+        action: 'handleGetPublicForm',
+        ...createReqMeta(req),
+        formId,
+      },
+      error,
+    })
+    const { errorMessage, statusCode } = mapRouteError(error)
+    return res.status(statusCode).json({ message: errorMessage })
+  }
+
+  const form = formResult.value
+  const publicFormView = form.getPublicView()
+  const { authType } = form
+
+  // NOTE: Once there is a valid form retrieved from the database,
+  // the client should always get a 200 response with the form's public view.
+  // Additional errors should be tagged onto the response object like myInfoError.
+  switch (authType) {
+    case AuthType.SP:
+    case AuthType.CP: {
+      // Form is valid, check for SPCP/MyInfo data.
+      return SpcpFactory.getSpcpSession(authType, req.cookies)
+        .map(({ userName }) =>
+          res.json({
+            form: publicFormView,
+            spcpSession: { userName },
+          }),
+        )
+        .mapErr((error) => {
+          // NOTE: Only log if there is no jwt present on the request.
+          // This is because clients can be members of the pubilc and hence, have no jwt.
+          if (!(error instanceof MissingJwtError)) {
+            logger.error({
+              message: 'Error getting public form',
+              meta: {
+                action: 'handleGetPublicForm',
+                ...createReqMeta(req),
+                formId,
+              },
+              error,
+            })
+          }
+          return res.json({
+            form: publicFormView,
+          })
+        })
+    }
+
+    case AuthType.MyInfo: {
+      const requestedAttributes = form.getUniqueMyInfoAttrs()
+
+      return (
+        // 1. Validate form and extract myInfoData
+        MyInfoFactory.extractMyInfoData(form, req.cookies)
+          // 2. Fill the form based on the result
+          .andThen((myInfoData) =>
+            MyInfoFactory.prefillMyInfoFields(
+              myInfoData,
+              form.toJSON().form_fields,
+            ).map((formFields) => ({
+              formFields,
+              spcpSession: { userName: myInfoData.getUinFin() },
+            })),
+          )
+          // 3. Hash and save to database
+          .andThen((form) =>
+            MyInfoFactory.saveMyInfoHashes(
+              form.spcpSession.userName,
+              formId,
+              form.formFields,
+            ).map(
+              // NOTE: Passthrough as form is needed in the pipeline
+              () => form,
+            ),
+          )
+          // 4. Return result if successful otherwise, clear cookies and return default response
+          .map(({ spcpSession, formFields }) =>
+            res.json({
```

Comment:
I think you might have missed out a couple of lines here which increment which check that `usedCount > 0` before retrieving the MyInfo data, and also increment `usedCount` in the cookie once the access token has been used. this is important because access tokens cannot be reused, so we need to track whether the tokens have been used before.